### ISSUE TITLE
`Remove-Item` reports incorrect number of removed items in the progress bar

### ISSUE THREAD

**Issue by kborowinski** [PowerShell/PowerShell#23875]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

1. Use `Remove-Item` on any folder with substantial number of **test files that you can delete.**
2. Cmdlet will report incorrect number of removed items in the progress bar.

It's either *off-by-one* on the total item count or on the currently removed item 🤷‍♂️

![obo-snapshot](https://github.com/PowerShell/PowerShell/assets/16168755/59ffaea5-6379-4090-ac96-4baed8919b49)

### Expected behavior

```console
Correct number of removed items should be reported in the progress bar
```

### Actual behavior

```console
Removed 4 of 3 files [...]
Removed 11 of 10 files [...]
```

### Error details

```console
No error
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.5.0-preview.3
PSEdition                      Core
GitCommitId                    7.5.0-preview.3-21-gbfa3dbe20db2758aa2abbbdd26166cf803893076
OS                             Microsoft Windows 10.0.22631
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0, 5.0, 5.1, 6.0, 7.0}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![obo](https://github.com/PowerShell/PowerShell/assets/16168755/3aa16784-3f45-4cec-9899-0fbc08f5ae65)

**Comment by jborean93** [comment]
This is just a guess but this could potentially be related to the whole race condition I mentioned in https://github.com/PowerShell/PowerShell/issues/23868#issuecomment-2137135760. The current code kicks off a task before it starts deleting called [GetTotalFiles](https://github.com/PowerShell/PowerShell/blob/bfa3dbe20db2758aa2abbbdd26166cf803893076/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L2884-L2888). The [GetTotalFiles](https://github.com/PowerShell/PowerShell/blob/bfa3dbe20db2758aa2abbbdd26166cf803893076/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L3643-L3682) function is in charge of calculating the total number of files and the total byte length and will st them on the `_totalFiles` and `_totalBytes` field as it encounters a new file. While this is running in the background the code will be deleting the files/folders and after each delete write the progress record based on the `_totalFiles` and `_totalBytes` that could still be adjusted by `GetTotalFiles` as it is running. https://github.com/PowerShell/PowerShell/blob/bfa3dbe20db2758aa2abbbdd26166cf803893076/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L3114-L3128

**Comment by MatejKafka** [comment]
This issue also results in the progress bar never being completed; if Remove-Item is invoked as part of a long-running script, the progress bar gets stuck at 100% and never disappears until the script finishes.

**Comment by kborowinski** [comment]
@jborean93 Is it easy to fix? Or it would require substantial code refactoring. The misleading progress bar info is quite annoying ;)

**Comment by jborean93** [comment]
I don’t think it would be an easy fix without impacting performance unfortunately.

**Comment by kborowinski** [comment]
If that's the case I wonder if the progress bar should be removed from the `Remove-Item` implementation. I don't think that it's OK to mislead users with incorrect information. What's you take on that @jborean93?

**Comment by jborean93** [comment]
I personally don't think the progress really adds much to the end user experience and just complicates the code even further but I can understand why others might disagree. I don't have any real sway/decision making in this project though so it'll be up to the UX working group to decide on.

**Comment by Bratman117** [comment]
Yeah I'm on 7.5.0-preview.4 and for some reason the progress bar stacks when calling Remove-Item multiple times in a script and is almost never correct.
![image](https://github.com/user-attachments/assets/91dfe610-6733-4994-94b4-d07c8ca3b7d0)
(My SSD is not that fast either...)

**Comment by corbob** [comment]
@tycho another option would be to add this to your profile or your scripts: `$PSDefaultParameterValues['Remove-Item:ProgressAction'] = 'SilentlyContinue'` This will set the `ProgressAction` only on `Remove-Item` only if it's not specifically passed in.

**Comment by jazzdelightsme** [comment]
@jborean93 I agree that it would be good to fix the underying race condition. But at a minimum, `Remove-Item` progress reporting should recognize when it gets "past 100%" and clean up anyway (the sort of thing you would add a telemetry assert for, to make sure you're not hiding new regressions, once the race condition gets fixed).

**Comment by adityapatwardhan** [comment]
The reason for the progress bar stacking is due to the following piece of code where it checks for removedItems to be equal to totalFiles to set the progress completion.

https://github.com/PowerShell/PowerShell/blob/21f88b50f468ef22fad0e0449ce7daa3f2aa4bd2/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L2953-L2960

**Comment by theJasonHelmick** [comment]
@kborowinski -- Thank you for the issue.  @jborean93 - you are correct, thank you for raising this point about the race condition. The WG is discussing how best to address this.

**Comment by corbob** [comment]
> The reason for the progress bar stacking is due to the following piece of code where it checks for removedItems to be equal to totalFiles to set the progress completion.
> 
> https://github.com/PowerShell/PowerShell/blob/21f88b50f468ef22fad0e0449ce7daa3f2aa4bd2/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L2953-L2960

So then am I being naive to think that changing the `==` to `>=` might resolve this?

**Comment by kborowinski** [comment]
Don't you dare to close it bot! 😏

**Comment by kborowinski** [comment]
> So then am I being naive to think that changing the `==` to `>=` might resolve this?

@corbob No, you're not 😉. I made a custom build of PowerShell, and this fixes the progress bar stacking. The off-by-one bug still remains, though. I've opened PR #27675 with this change. Even though it's not a complete fix, at least the stacking issue is resolved now!
